lightning_pose_app/ui/extract_frames.py

```diff
@@ -11,15 +11,116 @@
+
+from lightning_pose_app import (
+    LABELED_DATA_DIR,
+    MODELS_DIR,
+    MODEL_VIDEO_PREDS_INFER_DIR,
+    SELECTED_FRAMES_FILENAME,
+    VIDEOS_DIR,
+    VIDEOS_INFER_DIR,
+    VIDEOS_TMP_DIR,
+    ZIPPED_TMP_DIR,
+)
 from lightning_pose_app.utilities import StreamlitFrontend, abspath
 from lightning_pose_app.utilities import copy_and_reformat_video, get_frames_from_idxs
+from lightning_pose_app.utilities import compute_motion_energy_from_predection_df
 
 _logger = logging.getLogger('APP.EXTRACT_FRAMES')
 
 
+def identify_outliers(metrics, likelihood_thresh, thresh_metric_z):
+
+    # Initialize dictionary to store outlier flags for each metric
+    outliers = {m: None for m in metrics.keys()}
+    # Determine outliers for each metric
+    for metric, val in metrics.items():
+        if metric == "likelihood":
+            outliers[metric] = val < likelihood_thresh
+        else:
+            # Apply z-score and threshold to determine outliers
+            outliers[metric] = val.apply(zscore).abs() > thresh_metric_z
+    # Combine outlier flags from all metrics into a single 3D array
+    outliers_all = np.concatenate(
+        [d.to_numpy()[:, :, None] for _, d in outliers.items()],
+        axis=-1
+    )  # Shape: (n_frames, n_keypoints, n_metrics)
+    # Sum outlier flags to identify total outliers for each frame
+    outliers_total = np.sum(outliers_all, axis=(1, 2))
+    return outliers_total
+
+
+def run_kmeans(X, n_clusters):
+
+    kmeans_obj = KMeans(n_clusters, n_init="auto")
+    kmeans_obj.fit(X)
+    cluster_labels = kmeans_obj.labels_
+
+    return cluster_labels
+
+
+def select_max_frame_per_cluster(df):
+    # Copy the DataFrame to avoid modifying the original
+    df_copy = df.copy()
+    # Group by 'cluster_labels' and find the index of the max 'error score' in each group
+    idxs_max_error = df_copy.groupby('cluster_labels')['error score'].idxmax()
+    # Select the rows that correspond to the max 'error score' in each cluster
+    final_selection = df_copy.loc[idxs_max_error].sort_values(
+        by="frames index")["frames index"].tolist()
+
+    return final_selection
+
+
+def select_frames_using_metrics(preds,
+                                metrics,
+                                n_frames_per_video,
+                                likelihood_thresh,
+                                thresh_metric_z
+                                ):
+
+    me = compute_motion_energy_from_predection_df(preds, likelihood_thresh)
+    me_prctile = 50 if preds.shape[0] < 1e5 else 75  # take fewer frames if there are many
+    # Select index of high ME frames
+    idxs_high_me = np.where(me > np.percentile(me, me_prctile))[0]
+
+    # Store likelihood scores in metrics dictionary
+    kps_and_conf = preds.to_numpy().reshape(preds.shape[0], -1, 3)
+    conf = kps_and_conf[:, :, -1]
+    mask = preds.columns.get_level_values('coords').isin(['likelihood'])
+    metrics["likelihood"] = pd.DataFrame(conf)
```

Comment:
can remove this line, and lines 89 and 90; only need line 91 (compute likelihood mask) and line 94 (save likelihood)